### ISSUE TITLE
<stdatomic.h> is not yet supported when compiling as C, but this is planned for a future release

### ISSUE THREAD

**Issue by remia** [AcademySoftwareFoundation/openexr#1236]
While working on OCIO CI builds, I noticed this error I hadn't seen before. I saw working builds using the same version 3.1.4 before and the only difference I see so far is the compiler version MSVC 19.31.31104.0 ([failing](https://github.com/AcademySoftwareFoundation/OpenColorIO/runs/5352263794?check_suite_focus=true#step:9:100)) and MSVC 19.30.30709.0 ([working](https://github.com/AcademySoftwareFoundation/OpenColorIO/runs/5163192920?check_suite_focus=true#step:9:715)).

Our build script can be found [here](https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/share/ci/scripts/multi/install_openexr.sh) in case that helps.

**Comment by jespergustinmsft** [comment]
We bumped into the same issue here... Solved it by updating the files internal_structs.h and chunk.c to say:
```
#if defined __has_include
#    if __has_include(<stdatomic.h>)
#        if defined(_MSC_VER)
#            ifdef __cplusplus
#                define EXR_HAS_STD_ATOMICS 1
#            endif
#        else
#            define EXR_HAS_STD_ATOMICS 1
#        endif
#    endif
#endif
```
instead of:
```
#    if defined __has_include
#        if __has_include(<stdatomic.h>)
#            define EXR_HAS_STD_ATOMICS 1
#        endif
#    endif
```
to only set EXR_HAS_STD_ATOMICS to 1 if we're on msvc and C++.

Please note that this is probably not the correct solution, since _MSV_VER is actually not checked for its value (and probably other issues as well), but simply that it is defined. But I thought I would share it anyway, if you hadn't found any solution.

**Comment by remia** [comment]
Thanks @jespergustinmsft, actually we are not in any rush for these CI builds as these are our nightly "analysis" builds meant exactly for catching errors ahead of time.

Looking at `internal_structs.h` I have to say I'm not quite sure I understand how your changes works, `#ifdef __cplusplus` looks guaranteed to return false at that point. Did you mean something like this (c/c from `internal_structs.h`), note that the end result would be the same as your code I guess.

```
#ifdef __cplusplus
#    include <atomic>
using atomic_uintptr_t = std::atomic_uintptr_t;
#else
#    if !defined(_MSC_VER)
#        if defined __has_include
#            if __has_include(<stdatomic.h>)
#                define EXR_HAS_STD_ATOMICS 1
#            endif
#        endif
#    endif
```

**Comment by jespergustinmsft** [comment]
Oh! You're right of course. I didn't read the surrounding code close enough. Your solution is better! :) Thanks for pointing it out! And at some point it should be possible to make an even better fix that compares _MSC_VER > 193x, where the support is actually added to compile with the C compiler.

**Comment by remia** [comment]
Curious how that would compare to the versions I pasted above, that was suggesting newer version of MSVC was failing. I would have expected the inverse like you said! Or are you saying before 193x MSVC was not supporting C compiler?

**Comment by jespergustinmsft** [comment]
Yeah... I assume __has_include(<stdatomic.h>) must have been evaluated to false in 19.30 and true in 19.31 because partial (C++ only) support was added in 19.31, but no support at all in 19.30.

**Comment by ignus2** [comment]
I'd really like this to be fixed, basically OpenEXR doesn't compile out of the box with the latest Visual Studio 2022 currently.

**Comment by meshula** [comment]
I don't know if anyone on the core dev team here has installed VS2022 yet.

A PR would be most welcome if anyone on this thread has already got something that works for VS2022, 2019, and 2017.

**Comment by cary-ilm** [comment]
Closing, looks like this is resolved by #1237.
